bin/opd8n/src/cli.rs

```diff
@@ -11,25 +14,111 @@ pub struct Cli {
 }
 
+impl Cli {
+    /// Returns the verbosity level for the CLI
+    pub fn v(&self) -> u8 {
+        match &self.command {
+            Commands::FromL2 { global, .. } => global.v,
+            Commands::FromL1 { global, .. } => global.v,
+            Commands::Info { global, .. } => global.v,
+        }
+    }
+}
+
+/// Global Arguments for the CLI
+#[derive(Parser, Clone, Debug)]
+pub struct GlobalArgs {
+    /// Verbosity level (0-4)
+    #[arg(long, short, help = "Verbosity level (0-4)", action = ArgAction::Count)]
+    pub v: u8,
+}
+
 /// Subcommands for the CLI
 #[derive(Parser, Clone, Debug)]
 pub enum Commands {
     /// Loads L2 input from a file
     #[command(visible_alias = "l")]
-    Load {
+    FromL2 {
         /// Path to the L2 info file
         #[clap(short, long, help = "Path to the L2 block info file")]
         input: PathBuf,
+        /// Global arguments
+        #[clap(flatten)]
+        global: GlobalArgs,
+    },
+    /// Pulls in the batch calldata or blob data for a given L1 Block
+    FromL1 {
+        /// Arguments for the L1 command
+        #[command(flatten)]
+        l1_args: L1Args,
+        /// Global arguments
+        #[clap(flatten)]
+        global: GlobalArgs,
     },
-    // TODO: Add another subcommand that provides an interactive method for generating the
-    // derivation test fixtures
+    /// Gets the L2 block info including the l1 origin for the l2 block number.
+    #[command(visible_alias = "i")]
+    Info {
+        /// The L2 Chain ID
+        #[clap(long, help = "L2 chain ID")]
+        l2_chain_id: u64,
+        /// The L2 block number to get info for
+        #[clap(long, help = "L2 block number")]
+        l2_block: u64,
+        /// The rpc url to fetch L2 block info from.
+        #[clap(long, help = "RPC url to fetch L2 block info from")]
+        rpc_url: String,
+        /// Global arguments
+        #[clap(flatten)]
+        global: GlobalArgs,
+    },
+}
+
+/// CLI arguments for the `from-l1` subcommand of `opd8n`.
+#[derive(Parser, Clone, Debug)]
+pub struct L1Args {
+    /// The L1 block number to start from
+    #[clap(short, long, help = "Starting L1 block number")]
+    pub start_block: u64,
+    /// The L1 block number to end at
+    #[clap(short, long, help = "Ending L1 block number")]
+    pub end_block: u64,
+    /// A list of L2 output root hashes to assert against.
+    #[clap(short, long, help = "L2 output root hashes")]
+    pub outputs: Vec<B256>,
+    /// An RPC URL to fetch L1 block data from.
+    #[clap(long, help = "RPC url to fetch L1 block data from")]
+    pub rpc_url: String,
+    /// A beacon client to fetch blob data from.
+    #[clap(long, help = "Beacon client url to fetch blob data from")]
+    pub beacon_url: String,
+    /// Optionally derive l1 output roots by running derivation over
+    /// the provided l1 block range.
+    #[clap(long, help = "Derive L1 output roots")]
+    pub derive: bool,
+    /// The output file for the test fixture.
```

Comment:
Consider updating to a similar approach to `forge` where a struct is used for each set of subcommand args. Then the `run()` method can be implemented for each struct.

https://github.com/foundry-rs/foundry/blob/7c4482fc9541f11b57575e2d8bf7bd190b61bda6/crates/forge/bin/cmd/create.rs#L26-L152

Reply:
Great suggestion, I'll have these changes up asap. Thanks